atomic_scanners/misc-package-updates/scanner.py

```diff
@@ -0,0 +1,208 @@
+
+    outdated_packages = list_of_outdated_packages(response)
+
+    # if newer version of pip is being used inside the container under scan, a
+    # DEPRECATION warning is raised by pip; we must ignore this.
+
+    if "DEPRECATION:" in outdated_packages[0]:
+        return outdated_packages[1:]
+
+    return outdated_packages
+
+
+def npm_list_of_outdated_packages(response):
+    """
+    List of oudated packages for npm package manager
+    """
+    # like pip, npm returns '\n' separated output but first line can be ignored
+    response = split_by_newline(response)[1:]
+
+    response = remove_last_blank_character(response)
+
+    return list_of_outdated_packages(response)
+
+
+def gem_list_of_outdated_packages(response):
+    """
+    List of oudated packages for gem package manager
+    """
+    # like pip, gem returns '\n' separated output
+    response = split_by_newline(response)
+
+    response = remove_last_blank_character(response)
+
+    return list_of_outdated_packages(response)
+
+
+def format_response(cli_arg, response):
+    """
+    Based on the CLI argument provided, check for the package updates related to
+    the package manager
+    """
+    if cli_arg == 'pip':
+        return pip_list_of_outdated_packages(response)
+    elif cli_arg == "npm":
+        return npm_list_of_outdated_packages(response)
+    elif cli_arg == "gem":
+        return gem_list_of_outdated_packages(response)
+
+
+def template_json_data(scan_type):
+    """
+    Template data on top of which additional data that is provided by different
+    functions is added
+    """
+    current_time = datetime.now().strftime('%Y-%m-%d-%H-%M-%S-%f')
+    json_out = {
+        "Start Time": current_time,
+        "Successful": "",
+        "Scan Type": scan_type + "-updates",
+        "UUID": UUID,
+        "CVE Feed Last Updated": "NA",
+        "Scanner": "Misc Package Updates",
+        "Scan Results": {"{} package updates".format(cli_arg): []}
+    }
+    return json_out
+
+json_out = template_json_data(cli_arg)
+
+try:
+    # Create the container before starting/running it
+    container = client.create_container(image=IMAGE_NAME,
+                                        command="tail -f /dev/null")
+
+    # Running the container
+    client.start(container.get('Id'))
+
+    # Check for pip updates
+    if cli_arg == "pip":
+        # variable to store info about exec_start
+        exe = client.exec_create(
+            container=container.get("Id"),
+            cmd="pip list --outdated"
+        )
+
+        response = client.exec_start(exe)
+
+    # Check for rubygem updates
+    elif cli_arg == "gem":
+        exe = client.exec_create(
+            container=container.get("Id"),
+            cmd="gem outdated"
+        )
+
+        response = client.exec_start(exe)
+
+    # Check for npm updates
+    elif cli_arg == "npm":
+        exe = client.exec_create(
+            container=container.get("Id"),
+            cmd="npm outdated"
```

Comment:
@dharmit I am no NPM expert, so don't quote me on this :) I *think* if you want to check globally installed modules, you need to add `-g`, i.e.: `npm outdated -g`. Otherwise you're checking modules installed in `./npm_modules/` directory.